staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert.go

```diff
@@ -80,7 +80,7 @@ func NewStructural(s *apiextensions.JSONSchemaProps) (*Structural, error) {
 		ss.Items = item
 	}
 
-	if len(s.Properties) > 0 {
+	if s.Properties != nil {
```

Comment:
@liggitt are we fine with that?

Reply:
we could add a formal validation that `properties` for non-preserve-unknown-fields objects are never empty, making JSON and proto to behave the same.

Reply:
> @liggitt are we fine with that?

not as written.

we have to be **really** careful here... this could allow a user to submit the following data that would pass validation, then fail validation after round-tripping through etcd:
```yaml
    openAPIV3Schema:
      type: object
      properties:
        template:
          type: object
          x-kubernetes-embedded-resource: true
          properties: {}
```

in 1.15, even if `properties: {}` was submitted to the apiserver (via a manual JSON request), the `len(s.Properties) > 0` check here meant that the structural schema had a nil `Properties` field, and validation failed with `...properties: Required value: must not be empty if x-kubernetes-embedded-resource is true without x-kubernetes-preserve-unknown-fields` error

with this PR, the CRD is allowed to be created, and updates fail with that message (since round-tripping through protobuf in etcd truncated the properties field to nil).

Reply:
> we could add a formal validation that properties for non-preserve-unknown-fields objects are never empty, making JSON and proto to behave the same.

that seems best. we know no existing schemas or existing 1.15 API clients specified empty properties, because of the `len() > 0` check here.

I would:
* leave the `len() > 0` check here
* switch the `if s.XEmbeddedResource && !s.XPreserveUnknownFields && s.Properties == nil {` check to if `s.XEmbeddedResource && !s.XPreserveUnknownFields && len(s.Properties) == 0 {` to make the check match what we can round-trip

If someone really wants an embedded resource with no other properties without preserving unknown fields, they can explicitly define `apiVersion` / `kind` properties

Reply:
done ^^

Reply:
including tests.